HW2/P2/parallel_vector.pyx

```diff
@@ -51,53 +54,76 @@ cdef void free_N_locks(int N, omp_lock_t *locks) nogil:
 ##################################################
 
 cpdef move_data_serial(np.int32_t[:] counts,
-                       np.int32_t[:] src,
-                       np.int32_t[:] dest,
-                       int repeat):
-   cdef:
-       int idx, r
+                      np.int32_t[:] src,
+                      np.int32_t[:] dest,
+                      int repeat):
+  cdef:
+      int idx, r
 
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+  assert src.size == dest.size, "Sizes of src and dest arrays must match"
+  with nogil:
+      for r in range(repeat):
+          for idx in range(src.shape[0]):              
+              if counts[src[idx]] > 0:
+                  counts[dest[idx]] += 1
+                  counts[src[idx]] -= 1
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
-                             np.int32_t[:] src,
-                             np.int32_t[:] dest,
-                             int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
+                            np.int32_t[:] src,
+                            np.int32_t[:] dest,
+                            int repeat):
+  cdef:
+      int idx, r, bothEqual
+      omp_lock_t *locks = get_N_locks(counts.shape[0])
 
    ##########
    # Your code here
    # Use parallel.prange() and a lock for each element of counts to parallelize
    # data movement.  Be sure to avoid deadlock, and double-locking.
    ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
-
-   free_N_locks(counts.shape[0], locks)
+  with nogil:
+      for r in range(repeat):
+          
+          # Need to enforce some order to how/when we acquire a lock (for justification why look 
+          # at the Dining Philosophers problem) so we just arbitrarily chose that the source
+          # must be the larger of the two. Acquire the locks and release in the same order, and when
+          # the source and dest are equal, we only acquire the lock for one of them.
+          for idx in prange(src.shape[0], num_threads=4, schedule=dynamic):
+              bothEqual = 0
+              if src[idx] > dest[idx]:
+                acquire(&locks[src[idx]])
+                acquire(&locks[dest[idx]])
+              elif src[idx] < dest[idx]:
+                acquire(&locks[dest[idx]])
+                acquire(&locks[src[idx]])
+              elif src[idx] == dest[idx]:
+                acquire(&locks[src[idx]])
+                bothEqual = 1
+              if counts[src[idx]] > 0:
```

Comment:
Usually, you want to do a check again after acquiring the lock, since the condition might have changed between checking the condition and acquiring the lock